### PR TITLE
chore: enable -Werror=thread-safety and add missing annotations (part 2/2)

### PR DESCRIPTION
**Pull request by kostasrim**
Part 2/2. 

resolves fully #3448

* add missing annotations
* small mutex fixes
* enable -Werror=thread-safety for clang builds

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -20,6 +20,7 @@ endif()
 
 if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
     set(CMAKE_CXX_FLAGS "-Wthread-safety ${CMAKE_CXX_FLAGS}")
+    add_compile_options(-Werror=thread-safety)
```
Comment (kostasrim) [inline CMakeLists.txt:23]: From now on we enforce checks with `ABSL_GUARD`

**Comment by kostasrim** [comment]
https://github.com/dragonflydb/dragonfly/actions/runs/10618452784

File: src/server/common.h [hunk 2]
```diff
@@ -377,6 +377,30 @@ class ABSL_LOCKABLE ThreadLocalMutex {
   util::fb2::detail::FiberInterface* locked_fiber_{nullptr};
 };
 
+// Replacement of std::SharedLock that allows -Wthread-safety
+template <typename Mutex> class ABSL_SCOPED_LOCKABLE SharedLock {
```
Comment (dranikpg) [inline src/server/common.h:381]: but we'll follow up with moving that to helio, right?

**Review by dranikpg — APPROVED** [review]

File: src/server/common.h (hunk 2 above)
Comment (kostasrim) [reply src/server/common.h:381]: If Roman doesn't have any issues yes

**Pull request merged**